minimint-api/src/net/peers.rs

```diff
@@ -0,0 +1,47 @@
+use async_trait::async_trait;
+use minimint_api::PeerId;
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+
+/// Owned [`PeerConnections`] trait object type
+pub type AnyPeerConnections<M> = Box<dyn PeerConnections<M> + Send + Unpin + 'static>;
+
+/// Connection manager that tries to keep connections open to all peers
+///
+/// Production implementations of this trait have to ensure that:
+/// * Connections to peers are authenticated and encrypted
+/// * Messages are received exactly once and in the order they were sent
+/// * Connections are reopened when closed
+/// * Messages are cached in case of short-lived network interruptions and resent on reconnect, this
+///   avoids the need to rejoin the consensus, which is more tricky.
+///
+/// In case of longer term interruptions the message cache has to be dropped to avoid DoS attacks.
+/// The thus disconnected peer will need to rejoin the consensus at a later time.  
+#[async_trait]
+pub trait PeerConnections<T>
+where
+    T: Serialize + DeserializeOwned + Unpin + Send,
+{
+    /// Send a message to a specific peer.
+    ///
+    /// The message is sent immediately and cached if the peer is reachable and only cached
+    /// otherwise.
+    async fn send(&mut self, target: PeerId, msg: T);
+
+    /// Send a message to all peers
+    async fn send_all(&mut self, msg: T);
```

Comment:
Idk if having both these functions be part of the trait is optimal. At the very least there should be a default impl for `send_all` imo.

Alternatively we could `impl<'a> dyn PeerConnections + 'a` like we did for the `Database` trait:

https://github.com/fedimint/minimint/blob/ecabd5622fa9ab59b195715aaa23bc2f8efb8907/minimint-api/src/db/mod.rs#L37-L65